File: LeetCode/[M] - 1004. Max Consecutive Ones III/solution.go
package main

func longestOnes(nums []int, k int) int {
	mx, curr, l := 0, 0, 0
	zeroIdx := make([]int, 0, len(nums))
	zeroIdx = append(zeroIdx, -1)
	for i := 0; i < len(nums); i++ {
		if nums[i] == 0 {
			zeroIdx = append(zeroIdx, i)
			if k == 0 {
				if curr > mx {
					mx = curr
				}
				l++
			} else {
				k--
			}
		}
		curr = i - zeroIdx[l]
	}
	return max(mx, curr)
}

File: LeetCode/[M] - 1004. Max Consecutive Ones III/solution.py
class Solution:
    def longestOnes(self, nums: list[int], k: int) -> int:
        mx, curr, l = 0, 0, 0
        zero_idx = [-1]
        for i in range(len(nums)):
            if nums[i] == 0:
                zero_idx.append(i)
                if k == 0:
                    mx = max(mx, curr)
                    l += 1
                else:
                    k -= 1
            curr = i - zero_idx[l]
        return max(mx, curr)
